crates/uv-python/src/discovery.rs

```diff
@@ -1663,12 +1717,31 @@ impl FromStr for VersionRequest {
             Some(VersionRequest::MajorMinor(major, minor))
         }
 
+        fn extract_prerelease(s: &str) -> (&str, Option<Prerelease>) {
+            match Prerelease::from_str(s) {
+                Ok(prerelease) => match s.rsplit_once(&prerelease.kind.to_string()) {
```

Comment:
This seems pretty hacky to me, and I think isn't guaranteed to work. For example, `1.0b1+b1`. It's perhaps a contrived example, but it's a valid version string that will cause this code to choke. I think in general, the flow of parsing a full version to get the prerelease, throw away that version except for the pre-release and then try to re-parse the version is pretty tortured.

I don't know the original motivation for avoiding full `Version` parsing here, but how about this:

1. In this `FromStr` impl, just try to parse a full `Version`.
2. Get rid of the ad hoc `.` splitting.
3. Inspect the `Version` directly to determine which request to return.
4. For `VersionSpecifiers`, it looks like you're specifically only trying to parse the part of the version without a prelease? Is that right? If so, you can use the mutating methods on the parsed `Version` to strip what you don't want, serialize it to a string and then re-parse.

If `Version` fails to parse, then I think you can skip directly to trying to parse as a `VersionSpecifiers`.

Splitting some of this into helper functions might help.

Reply:
@BurntSushi, thanks for the excellent comment! So that you know- I've made changes following your guidelines. Hopefully, it is now more concise and readable. 
I don't know why it was avoided initially to construct the `Version` type. Perf?